src/typechecker/FStar.TypeChecker.Env.fs

```diff
@@ -842,38 +845,78 @@ let attrs_of_qninfo (qninfo : qninfo) : option<list<attribute>> =
 let lookup_attrs_of_lid env lid : option<list<attribute>> =
   attrs_of_qninfo <| lookup_qname env lid
 
-let fv_with_lid_has_attr env fv_lid attr_lid : bool =
+let fv_exists_and_has_attr env fv_lid attr_lid : bool * bool =
     match lookup_attrs_of_lid env fv_lid with
-    | None
-    | Some [] ->
-      false
+    | None ->
+      false, false
     | Some attrs ->
+      true,
       attrs |> BU.for_some (fun tm ->
          match (U.un_uinst tm).n with
          | Tm_fvar fv -> S.fv_eq_lid fv attr_lid
          | _ -> false)
 
+let fv_with_lid_has_attr env fv_lid attr_lid : bool =
+  snd (fv_exists_and_has_attr env fv_lid attr_lid)
+
 let fv_has_attr env fv attr_lid =
   fv_with_lid_has_attr env fv.fv_name.v attr_lid
 
+let cache_in_fv_tab (tab:BU.smap<'a>) (fv:fv) (f:unit -> (bool * 'a)) : 'a =
+  let s = (S.lid_of_fv fv).str in
+  match BU.smap_try_find tab s with
+  | None ->
+    let should_cache, res = f () in
+    if should_cache then BU.smap_add tab s res;
+    res
+
+  | Some r ->
+    r
+
+let type_is_erasable env fv =
+  let f () =
+     let ex, erasable = fv_exists_and_has_attr env fv.fv_name.v Const.erasable_attr in
+     ex,erasable
+     //unfortunately, treating the Const.must_erase_for_extraction_attr
+     //in the same way here as erasable_attr leads to regressions in fragile proofs,
+     //notably in FStar.ModifiesGen, since this expands the class of computation types
+     //that can be promoted from ghost to tot. That in turn results in slightly different
+     //smt encodings, leading to breakages. So, sadly, I'm not including must_erase_for_extraction
+     //here. In any case, must_erase_for_extraction is transitionary and should be removed
+  in
+  cache_in_fv_tab env.erasable_types_tab fv f
+
+let rec non_informative env t =
+    match (U.unrefine t).n with
+    | Tm_type _ -> true
+    | Tm_fvar fv ->
+      fv_eq_lid fv Const.unit_lid
+      || fv_eq_lid fv Const.squash_lid
+      || fv_eq_lid fv Const.erased_lid
+      || type_is_erasable env fv
+    | Tm_app(head, _) -> non_informative env head
+    | Tm_uinst (t, _) -> non_informative env t
+    | Tm_arrow(_, c) ->
+      (is_pure_or_ghost_comp c && non_informative env (comp_result c))
+      //It would be sound to also add this disjunct below,
+      //But, it leads to regressions in examples/rel/Benton2004.fst
+      //|| is_ghost_effect (comp_effect_name c)
```

Comment:
Is this (old) comment really true? Promoting `GTot (int -> int -> GTot int)` to `Tot (int -> int -> GTot int)` and being able to partially apply it seems bad?

Reply:
I think it is true. Both `int -> GTot int` and `int -> int -> GTot int` are erased to `unit` and any applications (even partial) are erased immediately to `()`. 

The comment is new as of my work on this feature. I would like to retain that disjunct ... but it ended up regressing code in tricky ways. Notably, in the SMT encoding, functions that were previously encoding as Tm_ghost_arrow now get encoded as Tm_arrow ...and at least in 1 brittle proof (Benton2004) this distinction seemed to matter. I hope to investigate this regression further and add this case to non-informative types.